### PR TITLE
feat(ssh): add banner to inform uses about device status

### PR DESCRIPTION
**Pull request by henrybarreto**

### CODE REVIEW

File: ssh/go.mod [hunk 1]
```diff
@@ -51,3 +51,5 @@ require (
 )
 
 replace github.com/shellhub-io/shellhub => ../
+
+replace github.com/gliderlabs/ssh => github.com/gustavosbarreto/ssh v1.0.0
```
Comment (otavio) [inline ssh/go.mod:55]: Remove the fork.

**Review by otavio — CHANGES_REQUESTED** [review]

File: ssh/go.mod (hunk 1 above)
Comment (henrybarreto) [reply ssh/go.mod]: Removed

File: pkg/api/internalclient/device.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func (c *client) DeviceLookup(lookup map[string]string) (*models.Device, []error
 	}
 
 	if resp.StatusCode() != http.StatusOK {
-		return nil, []error{err}
+		return nil, []error{errors.New("failed to get the device on the API")}
```
Comment (otavio) [inline pkg/api/internalclient/device.go:85]: Shouldn't `err` be included in the error?

File: ssh/session/session.go [hunk 3]
```diff
@@ -212,23 +136,31 @@ func newSession(ctx gliderssh.Context) (*Session, error) {
...
+
 	return session, nil
 }
 
 func (s *Session) checkFirewall() (bool, error) {
 	if err := s.api.FirewallEvaluate(s.Data.Lookup); err != nil {
+		defer log.WithError(err).WithFields(log.Fields{
+			"uid":   s.UID,
+			"sshid": s.SSHID,
+		}).Info("an error or a firewall rule block this connection")
```
Comment (otavio) [inline ssh/session/session.go:155]:
```suggestion
		}).Info("an error or a firewall rule blocked the connection")
```

File: ssh/session/session.go [hunk 4]
```diff
@@ -238,28 +170,24 @@ func (s *Session) checkFirewall() (bool, error) {
 func (s *Session) checkBilling() (bool, error) {
 	device, err := s.api.GetDevice(s.Device.UID)
 	if err != nil {
-		return false, errors.Join(ErrFindDevice, err)
-	}
+		defer log.WithError(err).WithFields(log.Fields{
+			"uid":   s.UID,
+			"sshid": s.SSHID,
+		}).Info("failed to get the device on billing evaluatation")
```
Comment (otavio) [inline ssh/session/session.go:176]:
```suggestion
		}).Info("failed to get the device on billing evaluation")
```

File: ssh/session/session.go [hunk 5]
```diff
@@ -238,28 +170,24 @@ func (s *Session) checkFirewall() (bool, error) {
...
-func (s *Session) dial(ctx gliderssh.Context, tunnel *httptunnel.Tunnel) (net.Conn, error) {
-	dialed, err := tunnel.Dial(ctx, s.Device.UID)
-	if err != nil {
-		return nil, errors.Join(ErrDial, err)
-	}
+	if evaluatation, status, _ := s.api.BillingEvaluate(device.TenantID); status != 402 && !evaluatation.CanConnect {
+		defer log.WithError(err).WithFields(log.Fields{
+			"uid":   s.UID,
+			"sshid": s.SSHID,
+		}).Info("an error or a billing rule block this connection")
```
Comment (otavio) [inline ssh/session/session.go:185]:
```suggestion
		}).Info("an error or a billing rule blocked the connection")
```

File: ssh/session/session.go [hunk 6]
```diff
@@ -294,12 +222,21 @@ func (s *Session) authenticate() error {
...
-func (s *Session) NewAgentConnection(config *gossh.ClientConfig) error {
+// connect connects the session's client to the session's agent.
+func (s *Session) connect(authOpt authFunc) error {
+	config := &gossh.ClientConfig{
+		User:            s.Target.Username,
+		HostKeyCallback: gossh.InsecureIgnoreHostKey(), // nolint: gosec
+	}
+
+	if err := authOpt(s, config); err != nil {
+		return errors.New("failed to generate the authentication information")
```
Comment (otavio) [inline ssh/session/session.go:233]:
```suggestion
		return errors.New("fail to generate the authentication information")
```

**Review by otavio — CHANGES_REQUESTED** [review]

**Comment by otavio** [comment]
Also, this is going to be a single commit, right?

File: pkg/api/internalclient/device.go (hunk 2 above)
Comment (henrybarreto) [reply pkg/api/internalclient/device.go:85]: In this case, the error in this arm is nil, that why I have created this hard-coded error.

Comment (otavio) [reply pkg/api/internalclient/device.go:85]:
```suggestion
		return nil, []error{errors.New("fail to get the device from the API")}
```

**Review by luannmoreira — APPROVED** [review]

File: ssh/server/server.go [hunk 7]
```diff
@@ -37,9 +40,65 @@ func NewServer(opts *Options, tunnel *httptunnel.Tunnel) *Server {
...
-		Addr:             ":2222",
-		PasswordHandler:  auth.PasswordHandler(tunnel),
-		PublicKeyHandler: auth.PublicKeyHandler(tunnel),
+		Addr: ":2222",
+		ConnCallback: func(ctx gliderssh.Context, conn net.Conn) net.Conn {
+			ctx.SetValue("conn", conn)
+
+			return conn
+		},
+		BannerHandler: func(ctx gliderssh.Context) string {
```
Comment (gustavosbarreto) [inline ssh/server/server.go:49]: You could optimize the repetitive `log.Fields` calls by creating the log object once and reusing it throughout the function.

Approach A:

```go
logFields := log.Fields{
    "uid":   ctx.SessionID(),
    "sshid": ctx.User(),
}

log.WithFields(logFields).Error("some error")
```

Approach B:

 ```
 ```go
logEntry := log.WithFields(log.Fields{
    "uid":   ctx.SessionID(),
    "sshid": ctx.User(),
})

logEntry.Error("some error")
```

Comment (otavio) [reply ssh/server/server.go:49]: I prefer `B` :-D

Comment (henrybarreto) [reply ssh/server/server.go:49]: Done

Comment (gustavosbarreto) [reply ssh/server/server.go:49]: > I prefer `B` :-D

Me too.

One advantage of Approach B is that it allows for the addition of more fields if necessary. By maintaining a base log entry for the entire function, and if needed, each entry can add more fields.

File: ssh/server/auth/password.go [hunk 8]
```diff
@@ -1,30 +1,48 @@
...
-				WithFields(log.Fields{"uid": ctx.SessionID()}).
-				Warn("failed to create a new session with password")
-
-			return false
+func PasswordHandler(ctx gliderssh.Context, passwd string) bool {
+	log.WithFields(
+		log.Fields{
+			"uid":   ctx.SessionID(),
+			"sshid": ctx.User(),
+		}).Trace("trying to use password authentication")
```
Comment (otavio) [inline ssh/server/auth/password.go:17]: Use a `logEntry` to avoid repeating.

File: ssh/server/auth/publickey.go [hunk 9]
```diff
@@ -1,30 +1,48 @@
...
-				WithFields(log.Fields{"uid": ctx.SessionID()}).
-				Warn("failed to create a new session with public key")
-
-			return false
+func PublicKeyHandler(ctx gliderssh.Context, publicKey gliderssh.PublicKey) bool {
+	log.WithFields(
+		log.Fields{
+			"uid":   ctx.SessionID(),
+			"sshid": ctx.User(),
+		}).Trace("trying to use public key authentication")
```
Comment (otavio) [inline ssh/server/auth/publickey.go:17]: Use a `logEntry` to avoid repeating.

**Review by otavio — CHANGES_REQUESTED** [review]

**Review by otavio — APPROVED** [review]

**Pull request merged**